### PR TITLE
HTTP API: support query the DDL history

### PR DESCRIPTION
**Pull request by ciscoxll**
support http api: query the all ddl history respond issues #6180.
curl ` http://{TiDBIP}:10080/ddl/history`
curl `http://{TiDBIP}:10080/ddl/history?limit={number}`

### CODE REVIEW

File: docs/tidb_http_api.md [hunk 1]
```diff
@@ -43,6 +43,11 @@
     ```shell
     curl http://{TiDBIP}:10080/schema/{db}
     ```
+1. Get All DDL History
```
Comment (jackysp) [inline docs/tidb_http_api.md]: Make it a sentence like others.

File: docs/tidb_http_api.md [hunk 2]
```diff
@@ -103,4 +108,4 @@ timezone.*
     ```shell
     curl http://{TiDBIP}:10080/tables/{colID}/{colFlag}/{colLen}?rowBin={val}
     ```
-    *Hint: For the column which field type is timezone dependent, e.g. `timestamp`, convert its value to UTC timezone.*
\ No newline at end of file
+    *Hint: For the column which field type is timezone dependent, e.g. `timestamp`, convert its value to UTC timezone.*
```
Comment (jackysp) [inline docs/tidb_http_api.md]: Why change this line?

File: server/http_handler.go [hunk 3]
```diff
@@ -294,6 +314,11 @@ type tableHandler struct {
 	op string
 }
 
+// histryHandler is the handler for list job histry.
+type histryHandler struct {
```
Comment (jackysp) [inline server/http_handler.go]: The struct name seems not right.

Comment (winkyao) [reply server/http_handler.go]: history?

File: server/http_handler.go [hunk 4]
```diff
@@ -273,6 +274,25 @@ func (t *tikvHandlerTool) handleMvccGetByHex(params map[string]string) (interfac
 	return t.getMvccByEncodedKey(encodedKey)
 }
 
+func (t *tikvHandlerTool) GetAllHistory() ([]*model.Job, error) {
```
Comment (jackysp) [inline server/http_handler.go]: Is there a way to not expose this function for test? I think the data in the test are known

File: server/http_handler.go [hunk 5]
```diff
@@ -294,6 +315,11 @@ type tableHandler struct {
 	op string
 }
 
+// historyHandler is the handler for list job histry.
+type historyHandler struct {
```
Comment (winkyao) [inline server/http_handler.go]: historyHandler is ambiguous, `ddlHistoryJobHandler` maybe better?

File: server/http_handler.go [hunk 6]
```diff
@@ -273,6 +274,26 @@ func (t *tikvHandlerTool) handleMvccGetByHex(params map[string]string) (interfac
...
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	store := domain.GetDomain(s.(sessionctx.Context)).Store()
+	txn, err := store.Begin()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	tt := meta.NewMeta(txn)
```
Comment (winkyao) [inline server/http_handler.go]: s/tt/txnMeta/

File: server/http_status.go [hunk 7]
```diff
@@ -54,6 +54,7 @@ func (s *Server) startHTTPServer() {
 	router.Handle("/schema/{db}", schemaHandler{tikvHandlerTool})
 	router.Handle("/schema/{db}/{table}", schemaHandler{tikvHandlerTool})
 	router.Handle("/tables/{colID}/{colTp}/{colFlag}/{colLen}", valueHandler{})
+	router.Handle("/ddl/history", historyHandler{tikvHandlerTool})
```
Comment (winkyao) [inline server/http_status.go]: Define a new handler like ddlHandler.

File: server/http_handler.go [hunk 8]
```diff
@@ -273,6 +274,26 @@ func (t *tikvHandlerTool) handleMvccGetByHex(params map[string]string) (interfac
 	return t.getMvccByEncodedKey(encodedKey)
 }
 
+func (t *tikvHandlerTool) GetAllHistoryDDL() ([]*model.Job, error) {
+	s, err := session.CreateSession(t.store.(kv.Storage))
```
Comment (shenli) [inline server/http_handler.go]: We should close the session.

Comment (ciscoxll) [reply server/http_handler.go]: Already close the session.

Comment (ciscoxll) [reply server/http_handler.go]: Already close the session.

Comment (shenli) [reply server/http_handler.go]: Got it.

**Comment by shenli** [comment]
LGTM

File: server/http_handler.go [hunk 9]
```diff
@@ -66,6 +67,7 @@ const (
 
 // For query string
 const qTableID = "table_id"
+const qLimitID = "limit"
```
Comment (lamxTyler) [inline server/http_handler.go]: `qLimitID` -> `qLimit`.

File: server/http_handler.go [hunk 10]
```diff
@@ -273,6 +275,28 @@ func (t *tikvHandlerTool) handleMvccGetByHex(params map[string]string) (interfac
 	return t.getMvccByEncodedKey(encodedKey)
 }
 
+func (t *tikvHandlerTool) GetAllHistoryDDL() ([]*model.Job, error) {
+	s, err := session.CreateSession(t.store.(kv.Storage))
+	defer s.Close()
```
Comment (lamxTyler) [inline server/http_handler.go]: If the `err` is not nil, then `s` will be `nil`, so this will cause panic?

File: server/http_handler.go [hunk 11]
```diff
@@ -294,6 +318,11 @@ type tableHandler struct {
 	op string
 }
 
+// ddlHistoryJobHandler is the handler for list job histry.
```
Comment (lamxTyler) [inline server/http_handler.go]: typo for `histry`.

File: server/http_handler_test.go [hunk 12]
```diff
@@ -543,3 +547,31 @@ func (ts *HTTPHandlerTestSuite) TestGetSchema(c *C) {
 	_, err = http.Get(fmt.Sprintf("http://127.0.0.1:10090/schema/tidb/abc"))
 	c.Assert(err, IsNil)
 }
+
+func (ts *HTTPHandlerTestSuite) TestAllHistory(c *C) {
+	ts.startServer(c)
+	ts.prepareData(c)
+	defer ts.stopServer(c)
+
+	resp, err := http.Get(fmt.Sprintf("http://127.0.0.1:10090/ddl/history/?limit=10"))
```
Comment (lamxTyler) [inline server/http_handler_test.go]: No need to `Sprintf`.

File: server/http_handler.go [hunk 13]
```diff
@@ -576,6 +605,41 @@ func (h tableHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+			return
+		}
+
+		if lid < 1 {
+			writeError(w, errors.New("ddl history limit must be greater than 1"))
+			return
+		}
+		jobs, err := h.GetAllHistoryDDL()
+		if err != nil {
+			writeError(w, errors.New("ddl history not found"))
```
Comment (lamxTyler) [inline server/http_handler.go]: Miss `return`?

File: server/http_handler.go [hunk 14]
```diff
@@ -576,6 +605,41 @@ func (h tableHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
+		}
+
+		writeData(w, jobs)
+		return
+	}
+	jobs, err := h.GetAllHistoryDDL()
+	if err != nil {
+		writeError(w, errors.New("ddl history not found"))
+	}
+	writeData(w, jobs)
```
Comment (lamxTyler) [inline server/http_handler.go]: If the `err` is not nil, do we still need to write data?

**Comment by ciscoxll** [comment]
@zimulala @winkyao PTAL

File: server/http_handler.go [hunk 15]
```diff
@@ -273,6 +275,31 @@ func (t *tikvHandlerTool) handleMvccGetByHex(params map[string]string) (interfac
 	return t.getMvccByEncodedKey(encodedKey)
 }
 
+func (t *tikvHandlerTool) GetAllHistoryDDL() ([]*model.Job, error) {
```
Comment (jackysp) [inline server/http_handler.go]: I think we do not need to expose this method.

File: server/http_status.go [hunk 16]
```diff
@@ -54,6 +54,8 @@ func (s *Server) startHTTPServer() {
 	router.Handle("/schema/{db}", schemaHandler{tikvHandlerTool})
 	router.Handle("/schema/{db}/{table}", schemaHandler{tikvHandlerTool})
 	router.Handle("/tables/{colID}/{colTp}/{colFlag}/{colLen}", valueHandler{})
+	router.Handle("/ddl/history", ddlHistoryJobHandler{tikvHandlerTool})
+	router.Handle("/ddl/history/{limit}/{number}", ddlHistoryJobHandler{tikvHandlerTool})
```
Comment (jackysp) [inline server/http_status.go]: It seems this handler is not used.

File: server/http_handler.go (hunk 15 above)
Comment (ciscoxll) [reply server/http_handler.go]: If we do not need to expose this method, then getting ddl job history is trouble.

**Pull request merged**